dg.go

```diff
@@ -1,315 +1,464 @@
+	if n.deleted {
+		return nil, &ErrNodeDeleted{n.id}
+	}
+	result := make(map[string]Node[NodeType], len(n.dg.connectionsFromNode[n.id]))
+	for toNodeID := range n.dg.connectionsFromNode[n.id] {
+		result[toNodeID] = n.dg.nodes[toNodeID]
+	}
+	return result, nil
 }
 
-func (d *node[NodeType]) Remove() error {
-    d.dg.lock.Lock()
-    defer d.dg.lock.Unlock()
-    if d.deleted {
-        return &ErrNodeDeleted{
-            d.id,
-        }
-    }
-    for toNodeID := range d.dg.connectionsFromNode[d.id] {
-        delete(d.dg.connectionsToNode[toNodeID], d.id)
-    }
-    delete(d.dg.connectionsFromNode, d.id)
-    for fromNodeID := range d.dg.connectionsToNode[d.id] {
-        delete(d.dg.connectionsFromNode[fromNodeID], d.id)
-    }
-    delete(d.dg.connectionsToNode, d.id)
-    delete(d.dg.nodes, d.id)
-    d.deleted = true
-    return nil
+// dependencyResolved is used to notify a node that one of its dependencies have had their resolution
+// status set. Once all dependencies are resolved, the node is set as finalized (ready for processing).
+// Caller should have appropriate mutex locked before calling.
+func (n *node[NodeType]) dependencyResolved(dependencyNodeID string, dependencyResolution ResolutionStatus) error {
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if dependencyResolution == Waiting {
+		// Illegal state
+		return ErrNotifiedOfWaiting{n.id, dependencyNodeID}
+	}
+	dependencyType, isOutstandingDependency := n.outstandingDependencies[dependencyNodeID]
+	if !isOutstandingDependency {
+		// Now determine if the missing item was because the dependency was already resolved, or
+		// because there was never a connection.
+		_, isConnected := n.dg.connectionsToNode[n.id][dependencyNodeID]
+		if isConnected {
+			// As designed, this is an internal function. So we guard against this in resolveNode.
+			panic(ErrDuplicateDependencyResolution{n.id, dependencyNodeID})
+		} else {
+			panic(ErrConnectionDoesNotExist{dependencyNodeID, n.id})
+		}
+	}
+	delete(n.outstandingDependencies, dependencyNodeID)
+	if dependencyType == ObviatedDependency {
+		return nil // Nothing to do.
+	}
+	// If resolution is unresolvable, fail if current type is AND, or if there are no remaining OR dependencies.
+	// Treat as resolved if is just a completion dependency.
+	if dependencyResolution == Unresolvable && dependencyType != CompletionDependency {
+		// Check for the failure case.
+		if dependencyType == AndDependency || !n.hasOutstandingDependency(OrDependency) {
+			// Missing requirement. Mark as unresolvable, which propagates to outbound connections.
+			n.ready = true
+			n.dg.readyForProcessing[n.id] = n
+			return n.resolveNode(Unresolvable)
+		}
+	} else {
+		var hasOrDependency bool
+		if dependencyType == OrDependency {
+			// No matter what, mark other ORs as obviated. If there are no AND dependencies left, mark as ready.
+			n.markOrsObviated()
+			hasOrDependency = false // This resolved all outstanding ORs.
+		} else {
```

Comment:
There is no code in this block which checks for AND dependencies, and there is no code which marks the node as ready, so I think you should remove the second sentence from the comment at line 430; then, without the second sentence, there is no need for the "No matter what" phrase; and that leaves us with "mark other ORs as obviated" which is pretty much just a restatement of the code, `n.markOrsObviated()`...so I suggest you remove the whole comment at line 430.  🙂

Reply:
You mean ... the obviation code has obviated its own comment?